src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs

```diff
@@ -2205,18 +2205,28 @@ internal Node RemoveAll(Predicate<T> match)
                 Contract.Ensures(Contract.Result<Node>() != null);
 
                 var result = this;
-                int index = 0;
-                foreach (var item in this)
+                var enumerator = new Enumerator(result);
+                try
                 {
-                    if (match(item))
+                    var startIndex = 0;
+                    while(enumerator.MoveNext())
                     {
-                        result = result.RemoveAt(index);
-                    }
-                    else
-                    {
-                        index++;
+                        if (match(enumerator.Current))
+                        {
+                            result = result.RemoveAt(startIndex);
+                            enumerator.Dispose();
```

Comment:
If the struct's Dispose method is indeed implemented so it is safe to call twice, then I have no concern and you don't have to worry about clearing the struct.
As for the cycling through the pool, that might actually be a cause for a small perf hit that would be nice if we could avoid. But again, I'm inclined to only perf tune when traces show it's worthwhile.